### PR TITLE
Support sparse vectors in LogisticRegression (Fit only)

### PR DESCRIPTION
**Pull request by lijinf2**

### CODE REVIEW

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/core.py [hunk 1]
```diff
@@ -313,7 +405,7 @@ def _pyspark_class(self) -> Optional[ABCMeta]:
         )
 
 
-class _CumlCaller(_CumlParams, _CumlCommon):
+class _CumlCaller(HasEnableSparseDataOptim, _CumlParams, _CumlCommon):
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/core.py:408]: This suggests we support sparse vectors in all our algos, while only logistic regression has support.   Seems it will break if sparse vectors are fed into other algos.

File: python/src/spark_rapids_ml/core.py [hunk 2]
```diff
@@ -552,11 +663,15 @@ def _call_cuml_fit_func(
         array_order = self._fit_array_order()
 
         cuml_verbose = self.cuml_params.get("verbose", False)
+        use_sparse_array = (
+            alias.featureVectorType in dataset.schema.fieldNames()
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/core.py:667]: Does this give a false positive if there is a column with the name alias.featureVectorType?   Maybe check specifically that the feature column is of struct type.

File: python/src/spark_rapids_ml/core.py [hunk 3]
```diff
@@ -135,6 +153,80 @@
...
+            # Note: According to spark ML VectorUDT format,
+            # when type field is 1, the size field is also empty.
+            # we need to check the values field to get vector length.
+            vec_size = len(vec_values)
+            csr_indices = np.arange(vec_size, dtype=np.int32)
+            csr_values = vec_values
+
+        if n_features == 0:
+            n_features = vec_size
+        assert n_features == vec_size
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/core.py:216]: Better to give more informative error message if this is not the case.  i.e all vectors must be of the same dimension.

File: python/src/spark_rapids_ml/core.py [hunk 4]
```diff
@@ -135,6 +153,80 @@
...
+
+        csr_indices_list.append(csr_indices)
+        csr_indptr_list.append(csr_indptr_list[-1] + len(csr_indices))
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/core.py:219]: Probably correct, but just checking it is expected that the length of this array is 1 more than the number of sparse rows.  ie. the last entry points to the start of the next non-existent row's index data.

File: python/src/spark_rapids_ml/params.py [hunk 5]
```diff
@@ -39,6 +39,28 @@
...
+    This is a Params based class inherited from XGBOOST: https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/spark/params.py.
+    It holds the variable to store the boolean config of enabling sparse data optimization.
+    """
+
+    enable_sparse_data_optim = Param(
+        Params._dummy(),
+        "enable_sparse_data_optim",
+        "This stores the boolean config of enabling sparse data optimization, if enabled, "
+        "Spark rapids ml will construct a sparse matrix in the the format of csr_matrix, "
+        "then calls cuml with the sparse matrix. This config is disabled by default. If most of "
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/params.py:54]: Technically it is enabled by default as the default None value bases sparse vs dense off first row.

File: python/src/spark_rapids_ml/utils.py [hunk 6]
```diff
@@ -444,3 +447,23 @@ def translate_trees(sc: SparkContext, impurity: str, model: Dict[str, Any]):  #
...
+    except ImportError:
+        pass
+
+    try:
+        from pyspark.databricks.sql.functions import unwrap_udt as databricks_unwrap_udt
+
+        return databricks_unwrap_udt
+    except ImportError as exc:
+        raise RuntimeError(
+            "Cannot import pyspark `unwrap_udt` function. Please install pyspark>=3.4 "
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/utils.py:467]: So thinking about our spark 3.3 test.   Let's document that this is only supported in spark >= 3.4 and disable those tests.

File: python/src/spark_rapids_ml/classification.py [hunk 7]
```diff
@@ -975,6 +989,7 @@ def _single_fit(init_parameters: Dict[str, Any]) -> Dict[str, Any]:
                     "n_cols": logistic_regression.n_cols,
                     "dtype": logistic_regression.dtype.name,
                     "num_iters": logistic_regression.solver_model.num_iters,
+                    "objective": logistic_regression.solver_model.objective,
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/classification.py:992]: Nice.   We can improve summary output with this.  I guess it is just final value not full history.

File: python/tests/test_logistic_regression.py [hunk 8]
```diff
@@ -1257,3 +1276,258 @@ def test_functor(y: np.ndarray, err_msg: str) -> None:
...
+    fit_intercept: bool,
+    caplog: LogCaptureFixture,
+) -> None:
+    datatype = np.float32
+    tolerance = 0.001
+    reg_param = 1e-6
+
+    from scipy.sparse import csr_matrix
+    from sklearn.datasets import fetch_20newsgroups
+    from sklearn.feature_extraction.text import CountVectorizer
```
Comment (eordentlich) [inline python/tests/test_logistic_regression.py:1426]: Would be nicer to use pyspark versions of CountVectorizer and train_test_split to add another example of api compatibility.

**Comment by lijinf2** [comment]
build

File: python/tests/test_logistic_regression.py (hunk 8 above)
Comment (eordentlich) [reply python/tests/test_logistic_regression.py:1428]: We should use this example to add sparse version of our example jupyter notebook: https://github.com/NVIDIA/spark-rapids-ml/blob/branch-23.12/notebooks/logistic-regression.ipynb

**Comment by lijinf2** [comment]
build

File: python/src/spark_rapids_ml/classification.py [hunk 9]
```diff
@@ -931,8 +939,14 @@ def _logistic_regression_fit(
                 concated = pd.concat(X_list)
                 concated_y = pd.concat(y_list)
             else:
-                # features are either cp or np arrays here
-                concated = _concat_and_free(X_list, order=array_order)
+                if isinstance(X_list[0], scipy.sparse.csr_matrix):
+                    concated = scipy.sparse.vstack(X_list)
+                elif isinstance(X_list[0], cupyx.scipy.sparse.csr_matrix):
+                    concated = cupyx.scipy.sparse.vstack(X_list)
```
[the remark this answers is not in the document]
Comment (lijinf2) [reply python/src/spark_rapids_ml/classification.py]: Done.

File: python/src/spark_rapids_ml/core.py [hunk 10]
```diff
@@ -135,6 +153,80 @@
...
+    unwrap_udt = _get_unwrap_udt_fn()
+    features_unwrapped_vec_col = unwrap_udt(feature_col)
+
+    # After a `pyspark.ml.linalg.VectorUDT` type column being unwrapped, it becomes
+    # a pyspark struct type column, the struct fields are:
+    #  - `type`: byte
+    #  - `size`: int
+    #  - `indices`: array<int>
+    #  - `values`: array<double>
+    # For sparse vector, `type` field is 0, `size` field means vector length,
```
[the remark this answers is not in the document]
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py]: Revised.

File: python/src/spark_rapids_ml/core.py [hunk 11]
```diff
@@ -379,17 +472,35 @@ def _pre_process_data(
...
+                        vector_to_array(col(input_col), vector_element_type).alias(alias.data)  # type: ignore
+                    )
+
                 if not self._float32_inputs:
                     feature_type = DoubleType
             else:
                 raise ValueError("Unsupported input type.")
 
-            dimension = len(dataset.first()[input_col])  # type: ignore
+            dimension = len(first_record[input_col])  # type: ignore
```
[the remark this answers is not in the document]
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py:509]: Yeah, see below copied from pyspark shell

>>> a = Vectors.sparse(10, [1., 2.], [1, 8])
>>> a
SparseVector(10, {1: 1.0, 2: 8.0})
>>> len(a)
10

File: python/src/spark_rapids_ml/core.py (hunk 1 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py]: Right. Moved the param to the algorithm class LogisticRegression.

File: python/src/spark_rapids_ml/core.py (hunk 2 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py:673]: It seems fine because core.py converts the feature column and label column into alias columns, through dataset = dataset.select(*select_cols). The select_cols all use alias names. 

Revised the code to also check the existence of alias.featureVectorSize and alias.featureVectorIndices, given the VectorUDT features column will be unwrapped into four alias columns (i.e. featureVectorType, featureVectorSize, featureVectorIndices, and data).

File: python/src/spark_rapids_ml/core.py (hunk 3 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py]: Added.

File: python/src/spark_rapids_ml/core.py (hunk 4 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py:219]: Added the checking of len(csr_indptr_list).

Did not add checking of last entry. It requires maintaining a counter to track the current number of non-zero values. Seems not checking is fine because the csr_matrix constructor used after will do some checking. Let me know if the checking is still favorable.

File: python/src/spark_rapids_ml/params.py (hunk 5 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/params.py]: Yeah, revised the docstring to be more accurate.

File: python/src/spark_rapids_ml/utils.py (hunk 6 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/utils.py:483]: Yeah, added the checking.

File: python/src/spark_rapids_ml/classification.py (hunk 7 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/classification.py:988]: Right, cuml currently only returns the final value. Need to update C++ side if to match the full summary

File: python/tests/test_logistic_regression.py (hunk 8 above)
Comment (lijinf2) [reply python/tests/test_logistic_regression.py]: Replaced the sklearn CountVectorizer with Spark CountVectorizer and RegexTokenizer. 

Thinking to add the example to notebook in the next PR. It seems the NLP20NEWS accuracy and training iterations are sensitive to the tokenizer and hyperparameters (i.e. RegParam and maxIter). May need to spend sometime to understand the behavior.

File: python/src/spark_rapids_ml/core.py [hunk 12]
```diff
@@ -135,6 +153,84 @@
...
+
+
+# similar to the XGBOOST _read_csr_matrix_from_unwrapped_spark_vec in https://github.com/dmlc/xgboost/blob/master/python-package/xgboost/spark/data.py
+def _read_csr_matrix_from_unwrapped_spark_vec(part: pd.DataFrame) -> csr_matrix:
+    # variables for constructing csr_matrix
+    csr_indices_list, csr_indptr_list, csr_values_list = [], [0], []
+
+    n_features = 0
+
+    for vec_type, vec_size_, vec_indices, vec_values in zip(
```
Comment (eordentlich) [inline python/src/spark_rapids_ml/core.py:194]: I wonder if there is a more efficient 'vectorized' approach to doing this.   Iterating in python can be slow.  Maybe add a TBD comment on this so we don't forget.

File: python/tests/test_logistic_regression.py (hunk 8 above)
Comment (eordentlich) [reply python/tests/test_logistic_regression.py]: Sure.  Another PR for notebook is fine.

File: python/src/spark_rapids_ml/core.py (hunk 12 above)
Comment (lijinf2) [reply python/src/spark_rapids_ml/core.py:195]: Yeah, added TBD comment.

**Comment by eordentlich** [comment]
I almost forgot about it, but what about sparse transform?   This PR seems to handle fit only.

**Comment by lijinf2** [comment]
build

**Comment by lijinf2** [comment]
> I almost forgot about it, but what about sparse transform? This PR seems to handle fit only.

Good catch. the PR forgets about transform. It runs but seems to convert everything to dense arrays. Will update.

**Review by eordentlich — APPROVED** [review]
Transform support can be added in a separate PR.
